src/server/data-sources/user-data-source.ts

```diff
@@ -0,0 +1,253 @@
+				},
+				(err, res) => {
+					if (err) {
+						const apolloError = new ApolloError(
+							'[UserDataSource] - An error occurred while trying to list users.',
+							ErrorCodes.LIST_USERS_FAILED,
+							err,
+						);
+						console.error(apolloError);
+						reject(apolloError);
+						return;
+					}
+
+					if (res.Items === undefined || res.Count === undefined || res.ScannedCount === undefined) {
+						const apolloError = new ApolloError(
+							'[UserDataSource] - The response of the "scan" command is invalid, thus operation to list users failed.',
+							ErrorCodes.INVALID_SCAN_RESPONSE,
+						);
+						console.error(apolloError);
+						reject(apolloError);
+						return;
+					}
+
+					console.log('[UserDataSource] - Successfully fetched list of users.');
+
+					resolve({
+						items: res.Items as UserModel[],
+						count: res.Count,
+						scannedCount: res.ScannedCount,
+						lastEvaluatedKey: res.LastEvaluatedKey,
+					});
+				},
+			);
+		});
+	}
+
+	async putItem(input: UserCreationModel): Promise<UserModel> {
+		return new Promise<UserModel>((resolve, reject) => {
+			console.log(`[UserDataSource] - Creating user with name "${input.name}".`);
+
+			// get current date as UTC to fulfill fields createdAt and updatedAt
+			const currentDate = new Date().toUTCString();
+
+			// fulfill some other props of the input, to make it ready to send to Dynamo
+			const item: UserModel = {
+				...input,
+				id: uuidv4(),
+				createdAt: currentDate,
+				updatedAt: currentDate,
+			};
+
+			// try to create user; if everything goes well, return the user; if the "put" command fails, catch, format and throw exception
+			this.docCient.put(
+				{
+					TableName: this.tableName,
+					Item: item,
+				},
+				(err, _res) => {
+					if (err) {
+						const apolloError = new ApolloError(
+							`[UserDataSource] - An error occurred while trying to create user with name ${item.name}.`,
+							ErrorCodes.PUT_USER_FAILED,
+							err,
+						);
+						console.error(apolloError);
+						reject(apolloError);
+						return;
+					}
+
+					console.log(`[UserDataSource] - Successfully created user with name "${item.name}".`);
+					resolve(item);
+				},
+			);
+		});
+	}
+
+	async updateItem(id: string, input: UserUpdateModel): Promise<UserModel> {
+		return new Promise<UserModel>((resolve, reject) => {
+			console.log(`[UserDataSource] - Updating user whose ID is "${id}".`);
+
+			// get current date as UTC to update the field updatedAt
+			const currentDate = new Date().toUTCString();
+
+			// generate the update object that will be passed to docCient.update
+			const updateItemInput = buildSimpleUpdateItemInput(this.tableName, id, {
+				...input,
+				updatedAt: currentDate,
+			});
+
+			// try to update user; if everything goes well, return the user; if the "update" command fails, catch, format and throw exception
+			this.docCient.update({ ...updateItemInput, ReturnValues: 'ALL_NEW' }, (err, res) => {
```

Comment:
Why did you choose to use callbacks instead of promises (async/await)?

Reply:
I actually was using promises before, but I ended up changing to callbacks just they were easier to mock with `jest`... I tried to use the package [jest-aws-sdk-mock](https://www.npmjs.com/package/jest-aws-sdk-mock), which would make things a lot easier to mock even with promises, but I could not make it work to mock lambdas ([known issue of the package](https://github.com/dwyl/aws-sdk-mock/issues/151)), so I decided to use purely `jest` instead and went back to using callbacks 😬